### PR TITLE
feat(dcutr): keep connection alive while we are using it

### PR DESCRIPTION
**Pull request by tcoratger**
## Description

Similar to #3876, we now compute `connection_keep_alive` based on whether we are still using the connection, applied to the `dcutr` protocol.

Related: https://github.com/libp2p/rust-libp2p/issues/3844.

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Review by thomaseizinger — COMMENTED** [review]
Thanks for tackling this!

The tests are failing because the handler has nothing to do, right after we established the connection.

We can fix this by passing in "to be done work" in the constructor of the handler.
Check the event handler for established connections in the associated behaviour. We are queuing an event there using `NotifyHandler` (https://github.com/libp2p/rust-libp2p/blob/cc5b346a7cde97573d96eef17904144cafb72de5/protocols/dcutr/src/behaviour_impl.rs#L131). Instead of passing that event to the handler via `NotifyHandler`, we should instead pass it in via the constructor of the handler.

That way, it immediately has something to do on startup and returns `KeepAlive::Yes`. Does that make sense?

**Comment by tcoratger** [comment]
> Thanks for tackling this!
> 
> The tests are failing because the handler has nothing to do, right after we established the connection.
> 
> We can fix this by passing in "to be done work" in the constructor of the handler. Check the event handler for established connections in the associated behaviour. We are queuing an event there using `NotifyHandler` (
> 
> https://github.com/libp2p/rust-libp2p/blob/cc5b346a7cde97573d96eef17904144cafb72de5/protocols/dcutr/src/behaviour_impl.rs#L131
> 
> ). Instead of passing that event to the handler via `NotifyHandler`, we should instead pass it in via the constructor of the handler.
> That way, it immediately has something to do on startup and returns `KeepAlive::Yes`. Does that make sense?

@thomaseizinger Yes I get your point here, so that you mean replacing

```rust
self.queued_events.extend([
    ToSwarm::NotifyHandler {
        peer_id,
        handler: NotifyHandler::One(connection_id),
        event: Either::Left(handler::relayed::Command::Connect {
            obs_addrs: self.observed_addreses(),
        }),
    },
    ToSwarm::GenerateEvent(Event::InitiatedDirectConnectionUpgrade {
        remote_peer_id: peer_id,
        local_relayed_addr: match connected_point {
            ConnectedPoint::Listener { local_addr, .. } => local_addr.clone(),
            ConnectedPoint::Dialer { .. } => unreachable!("Due to outer if."),
        },
    }),
]);
```
by something like (roughly):
```rust
handler::relayed::Handler::new(connected_point, true);
self.queued_events.extend([
    ToSwarm::GenerateEvent(Event::InitiatedDirectConnectionUpgrade {
        remote_peer_id: peer_id,
        local_relayed_addr: match connected_point {
            ConnectedPoint::Listener { local_addr, .. } => local_addr.clone(),
            ConnectedPoint::Dialer { .. } => unreachable!("Due to outer if."),
        },
    }),
]);
```
where `true` is a boolean value representing "to be done work". Then I add this condition into the `connection_keep_alive` to trigger the `KeepAlive::Yes` just after connection is established, right?

**Comment by thomaseizinger** [comment]
Almost! Instead of adding an extra boolean, have a look at what the `ConnectionHandler` does when it receives the message from the behaviour:

https://github.com/libp2p/rust-libp2p/blob/cc5b346a7cde97573d96eef17904144cafb72de5/protocols/dcutr/src/handler/relayed.rs#L287-L317

It adds an event to `queued_events`. If this queue is not empty, you are already returning `KeepAlive::Yes` in this PR! Thus, if you extend the constructor to take in the event itself and directly add it to the queue, all should be fine :)

Plus, this should allow us to remove the `Connect` message, making the implementation overall simpler!

File: protocols/dcutr/src/handler/relayed.rs [hunk 1]
```diff
@@ -310,14 +304,20 @@ impl ConnectionHandler for Handler {
                     );
                 }
             }
-            Command::UpgradeFinishedDontKeepAlive => {
-                self.keep_alive = KeepAlive::No;
-            }
+            Command::UpgradeFinishedDontKeepAlive => {}
```
Comment (thomaseizinger) [inline protocols/dcutr/src/handler/relayed.rs]: I think we can just delete this variant now?

**Comment by thomaseizinger** [comment]
> Plus, this should allow us to remove the `Connect` message, making the implementation overall simpler!

Unfortunately we can't because we still need it to trigger more attempts.

**Comment by thomaseizinger** [comment]
We do something similar in the relay itself already: https://github.com/libp2p/rust-libp2p/blob/cc5b346a7cde97573d96eef17904144cafb72de5/protocols/relay/src/priv_client.rs#L156-L174

File: protocols/dcutr/src/handler/relayed.rs (hunk 1 above)
Comment (tcoratger) [reply protocols/dcutr/src/handler/relayed.rs]: @thomaseizinger Yes you are right, this is done.

**Comment by thomaseizinger** [comment]
@tcoratger Did the [comment above](https://github.com/libp2p/rust-libp2p/pull/3960#issuecomment-1553001155) clear things up on how to proceed here?

**Comment by tcoratger** [comment]
> @tcoratger Did the [comment above](https://github.com/libp2p/rust-libp2p/pull/3960#issuecomment-1553001155) clear things up on how to proceed here?

@thomaseizinger Thank you for your explanations, I understood a little better but unfortunately, I do not yet have all the understanding of the nesting of functions on this subject, I must discover the library in more detail (I'm quite new on this).

1. I understood what must happen, i.e. when the `connection_keep_alive` function is called as soon as the connection is set up, then the `self.queued_events` must not be empty at the risk of returning `KeepAlive:: No`, which would turn off the connection when it shouldn't.

2. As with the `ToSwarm::NotifyHandler`, the event is already pushed to the handler, I still wonder why this basic implementation doesn't work: I guess because it's just a notification and the event does not have time to be pushed into the queue with the `on_behaviour_event` function, so `connection_keep_alive` reads an empty queue and therefore terminates the connection.

3. Based on this principle, I understand your approach which aims to directly integrate the event into the constructor of the handler. I thus propose the following or similar approach:
```rust 
// constructor with default values
 let mut handler = handler::relayed::Handler::new(connected_point.clone());
// push of the event (another approach can be adopted here)
  handler.on_behaviour_event(handler::relayed::Command::Connect {
      obs_addrs: self.observed_addresses(),
  });

  self.queued_events.extend([
      ToSwarm::GenerateEvent(Event::InitiatedDirectConnectionUpgrade {
          remote_peer_id: peer_id,
          local_relayed_addr: match connected_point {
              ConnectedPoint::Listener { local_addr, .. } => local_addr.clone(),
              ConnectedPoint::Dialer { .. } => unreachable!("Due to outer if."),
          },
      }),
  ]);
```
But using this approach, I don't understand how to push my `handler` thus created so that it appears in the queue of events transmitted from the behavior to the handler thereafter. The approach is not exactly the same as the one taken in `handle_established_inbound_connection` for the `relay` (for me) because here the `handler` is returned by the function and subsequently used to be pushed to the handler side I imagine.

In summary I have trouble seeing how to link the implementation of behavior with what is in the handler afterwards, probably because I lack experience in the library (that's why I try to choose themes that are not too complicated to begin to slowly familiarize myself with the codebase :)).

**Comment by thomaseizinger** [comment]
No worries at all, the life-cycles are tricky to understand!

I'll see to push some follow-up patches so you can see what I mean :)

**Comment by thomaseizinger** [comment]
@mxinden I pushed some follow-up commits here. The trickiest bit is that we need to keep the connection alive between the upgrade attempts. However, the probelab data showed that multiple upgrade attempts don't really increase the success rate.

We could simplify this a bit more if we remove the upgrade attempts. What do you think?

**Comment by thomaseizinger** [comment]
> @mxinden I pushed some follow-up commits here. The trickiest bit is that we need to keep the connection alive between the upgrade attempts. However, the probelab data showed that multiple upgrade attempts don't really increase the success rate.
> 
> We could simplify this a bit more if we remove the upgrade attempts. What do you think?

Another thing to discuss: Do we really need the events that notify us about a dcutr _attempt_? Those add code to the communication between handler and behaviour that is otherwise not needed.

**Comment by mxinden** [comment]
> Another thing to discuss: Do we really need the events that notify us about a dcutr _attempt_? Those add code to the communication between handler and behaviour that is otherwise not needed.

Initially I added them with the goal of exposing corresponding Prometheus metrics in `libp2p-metrics`. I think that is still a worthy goal.

**Comment by mxinden** [comment]
> @mxinden I pushed some follow-up commits here. The trickiest bit is that we need to keep the connection alive between the upgrade attempts. However, the probelab data showed that multiple upgrade attempts don't really increase the success rate.
> 
> We could simplify this a bit more if we remove the upgrade attempts. What do you think?

I would expect that we will re-introduce the retry logic in case we decide to implement https://github.com/libp2p/specs/issues/487. Thus my preference, unless they add significant amount of complexity, is to keep the retry logic.

File: protocols/dcutr/src/behaviour_impl.rs [hunk 2]
```diff
@@ -329,24 +326,17 @@ impl NetworkBehaviour for Behaviour {
...
-                            inbound_connect,
-                            obs_addrs: self.observed_addresses(),
-                        }),
-                    },
-                    ToSwarm::GenerateEvent(Event::RemoteInitiatedDirectConnectionUpgrade {
+            Either::Left(handler::relayed::Event::InboundConnectRequest) => {
+                self.queued_events.extend([ToSwarm::GenerateEvent(
+                    Event::RemoteInitiatedDirectConnectionUpgrade {
                         remote_peer_id: event_source,
-                        remote_relayed_addr: remote_addr,
```
Comment (mxinden) [inline protocols/dcutr/src/behaviour_impl.rs:341]: Why no longer provide the `remote_relayed_addr` with the `RemoteInitiatedDirectConnectionUpgrade` event? That would remove the need for explicit state tracking in the `NetworkBehaviour` implementation and instead store the state close to its source, namely the connection.

**Review by mxinden — COMMENTED** [review]
Thank you for the work here.

Shall we land https://github.com/libp2p/rust-libp2p/pull/3982 first?

**Comment by mxinden** [comment]
The merging of https://github.com/libp2p/rust-libp2p/pull/3982 introduced a tricky merge conflict here. @tcoratger I took a go at resolving it. Mind giving the latest merge commit a review?

**Comment by tcoratger** [comment]
> The merging of #3982 introduced a tricky merge conflict here. @tcoratger I took a go at resolving it. Mind giving the latest merge commit a review?

@mxinden Yes right, if I understand well your merge here:
- You have added, in the `handle_established_inbound_connection` function, the `handler` using the constructor and pushed that to the `queued_events` to mimic the previous pattern that was inside the match in order to establish connection.
- You put a `direct_connections` call in both `handle_established_inbound_connection` and `handle_established_outbound_connection` to push a direct and non relayed connection.

It sounds good to me, as I didn't code the most technical part of this PR, maybe @thomaseizinger can you take a look to see if everything is compatible with what you implemented?

File: protocols/dcutr/src/handler/relayed.rs [hunk 3]
```diff
@@ -286,38 +233,33 @@ impl ConnectionHandler for Handler {
...
-        self.keep_alive
+        if !self.queued_events.is_empty() {
+            return KeepAlive::Yes;
+        }
+
+        if self.inbound_connect.is_some() {
+            return KeepAlive::Yes;
+        }
+
+        if self.attempts < MAX_NUMBER_OF_UPGRADE_ATTEMPTS {
```
Comment (thomaseizinger) [inline protocols/dcutr/src/handler/relayed.rs:264]: Instead of referencing the constant here, I think it would be nicer if we passed the value in at the constructor.

That allows us to keep the constant an implementation detail of the behaviour. What do you think guys?

Comment (mxinden) [reply protocols/dcutr/src/handler/relayed.rs:264]: Passing it in would make the constant mutable, right? In other words, there is no way to define a constant in a `struct` in Rust. I find a global `const` easier to reason about than a maybe-const inside a `struct`.

File: protocols/dcutr/src/behaviour_impl.rs (hunk 2 above)
Comment (mxinden) [reply protocols/dcutr/src/behaviour_impl.rs:341]: Friendly ping. Am I missing something? This should eliminate the necessity for maintaining the state of Behaviour::remote_relayed_addr, correct?

**Review by mxinden — COMMENTED** [review]
Looks good to me overall. Nice to see these simplifications!

File: protocols/dcutr/src/handler/relayed.rs (hunk 3 above)
Comment (thomaseizinger) [reply protocols/dcutr/src/handler/relayed.rs:264]: It is mutable as a field then yes. At the moment though, the separation between behaviour and handler isn't very clean.

The handler trusts that the behaviour will operate on the same constant, it implicitly depends on it.

If we pass it in, the behaviour explicitly initializes the handler with this value, thus removing the dependency on the global const.

It makes the handler more cohesive which I'd argue is easier to understand. Mutating a variable that says "limit" would be a local bug that is pretty easy to find. Plus, it would be easy to write a unit test for this behaviour (not saying we need one).

File: protocols/dcutr/src/behaviour_impl.rs (hunk 2 above)
Comment (thomaseizinger) [reply protocols/dcutr/src/behaviour_impl.rs:341]: Yes-ish. The original idea was to not pass data back and forth. The behaviour learns about the address first so it is kind of redundant to pass it to the handler only to then pass it to the behaviour again.

On the flipside, not needing the hashmap removes a few error cases and simplifies the behaviour so I think it is a good idea.

@tcoratger Mind taking a look at this?

The idea would be to remove the `peer_addresses` hashmap and instead pass the address in the `InboundConnectRequest` enum from the handler to the behaviour.

Comment (tcoratger) [reply protocols/dcutr/src/behaviour_impl.rs:341]: @thomaseizinger @mxinden Done. I hope I have understood everything correctly.

From what I have in mind, the `peers_addresses` hashmap of the `Behavior` structure was used to list peer addresses. It was therefore updated when establishing or closing a connection.

To replace this and avoid some bug-prone cases, we prefer to simplify the `behavior` and place the logic in the `handler` instead. So I started by completely removing `peers_addresses` which is no longer useful. Then I added the `remote_addr` inside `InboundConnectRequest` to pass the address information from the handler, closer to the source of the connection.

Don't hesitate if there's something tricky that I didn't catch.

File: protocols/dcutr/src/handler/relayed.rs (hunk 3 above)
Comment (tcoratger) [reply protocols/dcutr/src/handler/relayed.rs:264]: Agreed with the logic here, since it is not possible to declare a constant inside a Rust structure, the best we could do is to "simulate" the behavior of a constant. So it's better for me to declare a real global constant.

**Review by mxinden — APPROVED** [review]
Looks good to me. Thank you for the follow-ups! @thomaseizinger let us know if you feel strongly about moving away from the `const`.

**Review by thomaseizinger — APPROVED** [review]
I think it would be cleaner the `const` being private to the behaviour but this is already a nice improvement overall so let's get it in!
